### PR TITLE
Gradle plugin update and Kotlin update

### PR DESCRIPTION
**Pull request by phyrian**
Tried to start using it in a Kotlin Multiplatform project, but found some missing implementations and failing plugin tasks, so fixed those. Hope it helps.

### CODE REVIEW

**Comment by ftomassetti** [comment]
Thank you! It is very much appreciated!

File: antlr-kotlin-examples-js/src/test/kotlin/MiscStringTest.kt [hunk 1]
```diff
@@ -11,19 +11,19 @@ fun assertArrayEquals(a: CharArray, b: CharArray) {
 class MiscStringTest {
 
     @Test fun testToCharArrayEmpty() {
-        assertArrayEquals(charArrayOf(), "".toCharArray())
+        assertArrayEquals(charArrayOf(), "".map { it }.toCharArray())
```
Comment (ftomassetti) [inline antlr-kotlin-examples-js/src/test/kotlin/MiscStringTest.kt]: Why this change? It is repeated a lot of times but I do not understand why

File: antlr-kotlin-examples-js/src/test/kotlin/MiscStringTest.kt [hunk 2]
```diff
@@ -11,19 +11,19 @@ fun assertArrayEquals(a: CharArray, b: CharArray) {
...
     }
 
     @Test fun testToCharArrayEmptyEl0() {
-        assertEquals('a', "abc def".toCharArray()[0])
+        assertEquals('a', "abc def".map { it }.toCharArray()[0])
     }
 
-    @Test fun testToCharArray() {
-        assertArrayEquals(charArrayOf('a', 'b', 'c', ' ', 'd', 'e', 'f'), "abc def".toCharArray())
+    @Test fun testmap { it }.toCharArray() {
```
Comment (ftomassetti) [inline antlr-kotlin-examples-js/src/test/kotlin/MiscStringTest.kt]: This is clearly wrong :)

File: antlr-kotlin-gradle-plugin/src/main/java/com/strumenta/antlrkotlin/gradleplugin/AntlrKotlinPlugin.java [hunk 3]
```diff
@@ -42,7 +42,7 @@
 public class AntlrKotlinPlugin implements Plugin<Project> {
     private static final Logger LOGGER = LoggerFactory.getLogger(AntlrKotlinPlugin.class);
 
-    private static final String ANTLR_CONFIGURATION_NAME = "antlrkotlin";
+    private static final String ANTLR_CONFIGURATION_NAME = "antlr-kotlin";
```
Comment (ftomassetti) [inline antlr-kotlin-gradle-plugin/src/main/java/com/strumenta/antlrkotlin/gradleplugin/AntlrKotlinPlugin.java]: Is this necessary?

File: antlr-kotlin-gradle-plugin/src/main/java/com/strumenta/antlrkotlin/gradleplugin/AntlrKotlinPlugin.java [hunk 4]
```diff
@@ -125,13 +125,20 @@ private void setupTask(final Project project) {
                     sourceSet.getJava().srcDir(outputDirectory);
 
                     // 6) register fact that antlr should be run before compiling
-                    Task compileKotlin = project.getTasks().findByName("compileKotlin");
-                    if (compileKotlin == null) {
-                        String message = "missing kotlin task, please add the gradle kotlin plugin to the build!";
+                    // compileKotlin is not available in the newer Multiplatform Kotlin plugins.
```
Comment (ftomassetti) [inline antlr-kotlin-gradle-plugin/src/main/java/com/strumenta/antlrkotlin/gradleplugin/AntlrKotlinPlugin.java]: Thank you! Making configuration more understandable was sorely needed

File: antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/ANTLRFileStream.kt [hunk 5]
```diff
@@ -5,23 +5,31 @@
...
-//        this.sourceName = fileName
-//        load(fileName, encoding)
-//    }
-//
-//    fun load(fileName: String, encoding: String?) {
-//        data = Utils.readFile(fileName, encoding)
-//        this.n = data.size
-//    }
-//}
+import com.soywiz.korio.lang.Charset
```
Comment (ftomassetti) [inline antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/ANTLRFileStream.kt]: These imports surprise me

File: antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/ANTLRInputStream.kt [hunk 6]
```diff
@@ -43,8 +42,9 @@ open class ANTLRInputStream : CharStream {
     constructor() {}
 
     /** Copy data in string to a local char array  */
+    
     constructor(input: String) {
-        this.data = input.toCharArray()
+        this.data = input.map { it }.toCharArray()
```
Comment (ftomassetti) [inline antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/ANTLRInputStream.kt]: Mmm, suspicious

File: antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/InterpreterRuleContext.kt [hunk 7]
```diff
@@ -20,10 +20,10 @@ package org.antlr.v4.kotlinruntime
  */
 class InterpreterRuleContext : ParserRuleContext {
     /** This is the backing field for [.getRuleIndex].  */
-    override var ruleIndex = -1
-        set(value: Int) {
-            super.ruleIndex = value
-        }
+//    override var ruleIndex = -1
+//        set(value: Int) {
+//            field = value
```
Comment (ftomassetti) [inline antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/InterpreterRuleContext.kt]: If it is not needed we should remove it

File: antlr-kotlin-examples-js/src/test/kotlin/MiscStringTest.kt (hunk 1 above)
Comment (phyrian) [reply antlr-kotlin-examples-js/src/test/kotlin/MiscStringTest.kt]: The String.toCharArray is currently experimental feature in kotlin common stdlib, but already there. Until they remove the experimental state of it, I think this is cleaner then providing the same method in the library, just to avoid collisions in naming.

File: antlr-kotlin-examples-js/src/test/kotlin/MiscStringTest.kt (hunk 2 above)
Comment (phyrian) [reply antlr-kotlin-examples-js/src/test/kotlin/MiscStringTest.kt]: Yes, I see, I accidentally replaced the methodname there too.

File: antlr-kotlin-gradle-plugin/src/main/java/com/strumenta/antlrkotlin/gradleplugin/AntlrKotlinPlugin.java (hunk 3 above)
Comment (phyrian) [reply antlr-kotlin-gradle-plugin/src/main/java/com/strumenta/antlrkotlin/gradleplugin/AntlrKotlinPlugin.java]: Not at all, but nicer. Does this affect anything outside?

File: antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/ANTLRFileStream.kt (hunk 5 above)
Comment (phyrian) [reply antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/ANTLRFileStream.kt]: There is a common lib by soywiz (a.k.a. korlibs) for IO utils, including multiplatform file reading, and the necessary struggles around it. Figured, might be good, since the code currently lacks a lot of IO features.

File: antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/ANTLRInputStream.kt (hunk 6 above)
Comment (phyrian) [reply antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/ANTLRInputStream.kt]: Collection<Char>.toCharArray() is defined by Kotlin Common Stdlib. String.toCharArray() is also defined but has experimental state.

File: antlr-kotlin-examples-js/src/test/kotlin/MiscStringTest.kt (hunk 2 above)
Comment (phyrian) [reply antlr-kotlin-examples-js/src/test/kotlin/MiscStringTest.kt]: Also, why are these added to examples?

**Comment by phyrian** [comment]
Found some issues with outputDirectory and packageName in Multiplatform projects, might be general as well. PackageName creates the wrong directory structure for the package and Antlr cannot find the generated .tokens file - doesn't look into the outputPackage correctly.

File: antlr-kotlin-examples-js/src/test/kotlin/MiscStringTest.kt (hunk 1 above)
Comment (ftomassetti) [reply antlr-kotlin-examples-js/src/test/kotlin/MiscStringTest.kt]: I would suggest we instead create an helper method (that we can also inline), so that we avoid repeating this pattern over and over. I think it would make for more readable code.

For example:

```
inline fun String.asCharArray() = this.map { it }.toCharArray()

...
assertArrayEquals(charArrayOf(), "".asCharArray())
...

```

Comment (phyrian) [reply antlr-kotlin-examples-js/src/test/kotlin/MiscStringTest.kt]: I can agree on that

File: antlr-kotlin-examples-jvm/src/test/kotlin/com/strumenta/antlrkotlin/examples/MiscStringTest.kt [hunk 8]
```diff
@@ -14,21 +14,21 @@ fun assertArrayEquals(a: CharArray, b: CharArray) {
...
 
     @Test
     fun testToCharArrayEmptyEl0() {
-        assertEquals('a', "abc def".toCharArray()[0])
+        assertEquals('a', "abc def".map { it }.toCharArray()[0])
     }
 
-//    @Test fun testToCharArray() {
-//        assertArrayEquals(charArrayOf('a', 'b', 'c', ' ', 'd', 'e', 'f'), "abc def".toCharArray())
+//    @Test fun testmap { it }.toCharArray() {
```
Comment (ftomassetti) [inline antlr-kotlin-examples-jvm/src/test/kotlin/com/strumenta/antlrkotlin/examples/MiscStringTest.kt]: We should not change this

File: antlr-kotlin-gradle-plugin/src/main/java/com/strumenta/antlrkotlin/gradleplugin/AntlrKotlinPlugin.java (hunk 3 above)
Comment (ftomassetti) [reply antlr-kotlin-gradle-plugin/src/main/java/com/strumenta/antlrkotlin/gradleplugin/AntlrKotlinPlugin.java]: Ok, makes sense

File: antlr-kotlin-gradle-plugin/src/main/java/com/strumenta/antlrkotlin/gradleplugin/AntlrKotlinTask.java [hunk 9]
```diff
@@ -254,21 +242,24 @@ private void evaluate(AntlrResult result) {
...
-     * flattened in the specified output directory.
-     *
-     * @param source The source.
-     * @since 4.0
-     */
-    @Override
-    public void setSource(FileTree source) {
-        setSource((Object) source);
-    }
+// Remove this: this is just simply unnecessary, it overwrites the default setter operator
```
Comment (ftomassetti) [inline antlr-kotlin-gradle-plugin/src/main/java/com/strumenta/antlrkotlin/gradleplugin/AntlrKotlinTask.java]: Please remove it then, instead of commenting it out

File: antlr-kotlin-runtime/build.gradle.kts [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+kotlin {
+    jvm()
+    js {
+        browser {
+        }
+        nodejs {
+        }
+    }
+    // For ARM, should be changed to iosArm32 or iosArm64
```
Comment (ftomassetti) [inline antlr-kotlin-runtime/build.gradle.kts]: We should add this explanation also in the Readme, I think

File: antlr-kotlin-runtime/build.gradle.kts [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+            dependencies {
+                implementation(kotlin("stdlib-js"))
+            }
+        }
+        val jsTest by getting {
+            dependencies {
+                implementation(kotlin("test-js"))
+            }
+        }
+//        val mingwMain by getting {
```
Comment (ftomassetti) [inline antlr-kotlin-runtime/build.gradle.kts]: Can we remove these?

File: antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/ANTLRFileStream.kt (hunk 5 above)
Comment (ftomassetti) [reply antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/ANTLRFileStream.kt]: Ok, good

File: antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/ANTLRFileStream.kt [hunk 12]
```diff
@@ -5,23 +5,31 @@
...
+import com.soywiz.korio.lang.UTF8
+import org.antlr.v4.kotlinruntime.misc.Utils
+
+/**
+ * This is an [ANTLRInputStream] that is loaded from a file all at once
+ * when you construct the object.
+ *
+ */
+@Deprecated("as of 4.7 Please use {@link CharStreams} interface.")
+
```
Comment (ftomassetti) [inline antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/ANTLRFileStream.kt]: I would remove this empty line

File: antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/ANTLRFileStream.kt [hunk 13]
```diff
@@ -5,23 +5,31 @@
...
-//    }
-//}
+import com.soywiz.korio.lang.Charset
+import com.soywiz.korio.lang.UTF8
+import org.antlr.v4.kotlinruntime.misc.Utils
+
+/**
+ * This is an [ANTLRInputStream] that is loaded from a file all at once
+ * when you construct the object.
+ *
```
Comment (ftomassetti) [inline antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/ANTLRFileStream.kt]: I would remove this empty line

File: antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/ANTLRInputStream.kt (hunk 6 above)
Comment (ftomassetti) [reply antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/ANTLRInputStream.kt]: please reuse the extension method we discussed

File: antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/Lexer.kt [hunk 14]
```diff
@@ -318,7 +316,7 @@ abstract class Lexer : Recognizer<Int, LexerATNSimulator>, TokenSource {
 
     fun getErrorDisplay(s: String): String {
         val buf = StringBuilder()
-        for (c in s.toCharArray()) {
+        for (c in s.map { it }.toCharArray()) {
```
Comment (ftomassetti) [inline antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/Lexer.kt]: Please reuse the utility method also here

File: antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/Recognizer.kt [hunk 15]
```diff
@@ -74,24 +71,24 @@ abstract class Recognizer<Symbol, ATNInterpreter : ATNSimulator> {
         get() {
             val vocabulary = vocabulary
             synchronized(tokenTypeMapCache) {
-                var result: MutableMap<String, Int>? = tokenTypeMapCache[vocabulary]
+                var result: Map<String, Int>? = tokenTypeMapCache[vocabulary]
                 if (result == null) {
-                    result = HashMap()
+                    val tmp = HashMap<String, Int>()
```
Comment (ftomassetti) [inline antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/Recognizer.kt]: `tmp` is not a great name. Can we think of some other name?

File: antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/atn/PredictionContext.kt [hunk 16]
```diff
@@ -211,37 +208,37 @@ abstract class PredictionContext protected constructor(
                 rootIsWildcard: Boolean,
                 mergeCache: DoubleKeyMap<PredictionContext, PredictionContext, PredictionContext>?
         ): PredictionContext {
-            var a = a
-            var b = b
-            assert(a != null && b != null) // must be empty context, never null
+            var a1 = a
```
Comment (ftomassetti) [inline antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/atn/PredictionContext.kt]: `a1` and `b1` do not convince me much as names

File: antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/misc/IntegerList.kt [hunk 17]
```diff
@@ -267,11 +267,11 @@ open class IntegerList {
         }
 
         val originalArray = _data!!.toTypedArray()
-        val copiedArray = Arrays.copyOf(originalArray, newLength)
+        val copiedArray = originalArray.copyOf(newLength)
         for (i in originalArray.size until copiedArray.size) {
             copiedArray[i] = 0
         }
-        _data = copiedArray.toIntArray()
+        _data = copiedArray.filterNotNull().toIntArray()
```
Comment (ftomassetti) [inline antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/misc/IntegerList.kt]: mmm, is this intended?

File: antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/misc/Utils.kt [hunk 18]
```diff
@@ -50,7 +29,7 @@ object Utils {
 
     fun escapeWhitespace(s: String, escapeSpaces: Boolean): String {
         val buf = StringBuilder()
-        for (c in s.toCharArray()) {
+        for (c in s.map { it }.toCharArray()) {
```
Comment (ftomassetti) [inline antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/misc/Utils.kt]: please use the utils method we discussed

File: antlr-kotlin-runtime/src/commonTest/kotlin/StringTest.kt [hunk 19]
```diff
@@ -1,26 +1,26 @@
-import com.strumenta.kotlinmultiplatform.toCharArray
 import kotlin.test.Test as test
 
+
 class MiscStringTest : BaseTest() {
 
     @test
     fun testToCharArrayEmpty() {
-        assertArrayEquals(charArrayOf(), "".toCharArray())
+        assertArrayEquals(charArrayOf(), "".map { it }.toCharArray())
```
Comment (ftomassetti) [inline antlr-kotlin-runtime/src/commonTest/kotlin/StringTest.kt]: please use the utility method we discussed

File: antlr-kotlin-target/src/main/resources/org/antlr/v4/tool/templates/codegen/Kotlin/Kotlin.stg [hunk 20]
```diff
@@ -939,7 +935,7 @@ private val serializedATN : String = Utils.join(
 private const val serializedATN : String = "<model.serialized>"
 
 <endif>
-val ATN = ATNDeserializer().deserialize(serializedATN.toCharArray())
+val ATN = ATNDeserializer().deserialize(serializedATN.map { it }.toCharArray())
```
Comment (ftomassetti) [inline antlr-kotlin-target/src/main/resources/org/antlr/v4/tool/templates/codegen/Kotlin/Kotlin.stg]: please use the utility method we discussed

File: build.gradle.kts [hunk 21]
```diff
@@ -7,7 +7,7 @@ buildscript {
     }
 
     dependencies {
-        classpath("org.jetbrains.kotlin:kotlin-gradle-plugin:$embeddedKotlinVersion")
+        classpath("org.jetbrains.kotlin:kotlin-gradle-plugin:1.3.50")
```
Comment (ftomassetti) [inline build.gradle.kts]: Can we use a variable or a property for the version number?

File: settings.gradle.kts [hunk 22]
```diff
@@ -5,3 +5,5 @@ include(
         "antlr-kotlin-target",
         "antlr-kotlin-gradle-plugin"
 )
+
+enableFeaturePreview("GRADLE_METADATA")
```
Comment (ftomassetti) [inline settings.gradle.kts]: Perhaps we can add a comment about this feature

File: antlr-kotlin-examples-jvm/src/test/kotlin/com/strumenta/antlrkotlin/examples/MiscStringTest.kt (hunk 8 above)
Comment (phyrian) [reply antlr-kotlin-examples-jvm/src/test/kotlin/com/strumenta/antlrkotlin/examples/MiscStringTest.kt]: Fixed, still WIP

File: antlr-kotlin-gradle-plugin/src/main/java/com/strumenta/antlrkotlin/gradleplugin/AntlrKotlinTask.java (hunk 9 above)
Comment (phyrian) [reply antlr-kotlin-gradle-plugin/src/main/java/com/strumenta/antlrkotlin/gradleplugin/AntlrKotlinTask.java]: Figured it had a purpose, so removed the comment

File: antlr-kotlin-runtime/build.gradle.kts (hunk 10 above)
Comment (phyrian) [reply antlr-kotlin-runtime/build.gradle.kts]: That comment autogenerates with every multiplatform lib. Currently I haven't done the native target, so it's not yet reasoned

File: antlr-kotlin-runtime/build.gradle.kts (hunk 11 above)
Comment (phyrian) [reply antlr-kotlin-runtime/build.gradle.kts]: Going to add TODO there, so it shows the intention

File: antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/misc/IntegerList.kt (hunk 17 above)
Comment (phyrian) [reply antlr-kotlin-runtime/src/commonMain/kotlin/org/antlr/v4/kotlinruntime/misc/IntegerList.kt]: Unfortunately there are many things in the code where we use nullable types. This is one of them. We should review those if it's really nullable or not. If it can be casted to Collection<Int> it's most likely shouldn't be Collection<Int?> in the first place.

**Comment by ftomassetti** [comment]
I will take another look when the build is fixed

File: antlr-kotlin-examples-js/build.gradle [hunk 23]
```diff
@@ -1,6 +1,6 @@
 buildscript {
-    ext.kotlin_version = '1.2.60'
-    ext.projectVersion = '0.0.2'
+    ext.kotlin_version = '1.3.50'
+    ext.projectVersion = '0.0.5'
```
Comment (ftomassetti) [inline antlr-kotlin-examples-js/build.gradle]: well, we could set it to version 0.1.0

File: antlr-kotlin-gradle-plugin/src/main/java/com/strumenta/antlrkotlin/gradleplugin/AntlrKotlinPlugin.java [hunk 24]
```diff
@@ -74,9 +74,11 @@ private void setupClasspath(final Project project) {
             defaultDependencies.add(projectDependencies
                     .create("com.strumenta.antlr-kotlin:antlr-kotlin-target:" + antlrKotlinVersion));
 
-            LOGGER.info("using antlr-kotlin-runtime-{} version {}", platform, antlrKotlinVersion);
-            defaultDependencies.add(projectDependencies
-                    .create("com.strumenta.antlr-kotlin:antlr-kotlin-runtime-" + platform + ":" + antlrKotlinVersion));
+            // FIXME: removed it until I figure out how to make it work for Multiplatform project.
```
Comment (ftomassetti) [inline antlr-kotlin-gradle-plugin/src/main/java/com/strumenta/antlrkotlin/gradleplugin/AntlrKotlinPlugin.java]: At the very least we should remember to open an issue for this. To me it seems an issue not having the dependency

File: antlr-kotlin-runtime/build.gradle.kts [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+            dependencies {
+                implementation(kotlin("stdlib-js"))
+            }
+        }
+        val jsTest by getting {
+            dependencies {
+                implementation(kotlin("test-js"))
+            }
+        }
+        /* TODO: Add Native target
```
Comment (ftomassetti) [inline antlr-kotlin-runtime/build.gradle.kts]: We should probably remove this

Comment (phyrian) [reply antlr-kotlin-runtime/build.gradle.kts]: We can move this out to an issue for sure, until then I want to have a reminder

**Comment by ftomassetti** [comment]
This is an amazing contribution! Thank you @phyrian for your work!
